### ISSUE TITLE
RC17 contains a heap memory leak?

### ISSUE THREAD

**Issue by Fristi** [zio/zio#2299]
Thanks for all the hard work being put in ZIO, the ecosystem starts to grow in something really great.

Despite that, we are seeing an issue with the latest version. We upgraded a few days ago to RC17. I think the major change is the auto supervision right? Since that release we see the heap memory increasing and eventual our app crashes due OOM.

<img width="409" alt="Screenshot 2019-11-21 at 11 45 55" src="https://user-images.githubusercontent.com/141265/69331184-a13ae880-0c54-11ea-88ae-ae0f8f6b291e.png">

We use
- ZIO RC17
- ZIO cats interop RC8
- fs2-kafka latest version

In this app we consume and produce messages from/to Kafka. So question is, do you have any idea what's causing this problem? Could we provide more information?

Also on a critical note, while there are a lot of additions and improvements to ZIO is there are any scope freeze being planned? It's hard to make something stable when there are still so many additions and changes to the current package. It's already RC17 for 1.0.0 and ZIO is being advertised to be production worthy while in some corner cases it might cause problems.

**Comment by neko-kai** [comment]
@jdegoes @adamgfraser @ghostdogpr 
All interop-cats fibers are created in a `.nonDaemon` region. Could this cause them to be added to root fibers set (if we have such a thing, i didn't look in detail) and leak memory in that way?

**Comment by neko-kai** [comment]
@Fristi Could you wrap your entire app in `.nonDaemon` and check if that changes the memory picture?

**Comment by jdegoes** [comment]
@neko-kai They will be added to the root fiber set. That can leak memory if the fibers never finish.

@Fristi The auto-supervision was the last feature we wanted to add before the 1.0 freeze. Unfortunately it's almost certainly the cause of the current issue, although it's hard to tell yet if it's a problem in ZIO or in user-land code.

**Comment by jdegoes** [comment]
@neko-kai Sorry, only `daemon` fibers will be added to the root set; the others will be added to their parent. The only way to leak in either case (in theory) is if the fibers never finish.

**Comment by Fristi** [comment]
Thanks for the quick response @neko-kai and @jdegoes. Nice to hear that current package is almost ready for a stable release 👍 🎉 

We'll try to add `.nonDaemon` around the involved fs2-kafka code to see if that helps. We'll get back to you as soon we know more.

> Unfortunately it's almost certainly the cause of the current issue, although it's hard to tell yet if it's a problem in ZIO or in user-land code.

Correct, if that does not help we will try to come up with a app which reproduces the issue.

**Comment by Fristi** [comment]
Would also be good to document `nonDaemon` this in the interop-cats documentation page I think. If it helps I'll PR something about that later

**Comment by Fristi** [comment]
We deployed a the `nonDaemon` change yesterday and we ran our endurance test for a bit. It seems it still suffers from the memory leak problem. We want to come up with a reproducible test case, to reach a case we are stripping some logic and see if we can pin point the problem.  @sebver will try to pin point and might get back to you later next week on that. I'll be offline for a week.

**Comment by neko-kai** [comment]
@sebver Any luck with a reproduction yet?

**Comment by neko-kai** [comment]
@jdegoes 
> @neko-kai Sorry, only daemon fibers will be added to the root set; the others will be added to their parent. The only way to leak in either case (in theory) is if the fibers never finish.

This is most likely the cause of the problem here. There's some code in fs2 that starts-and-forgets fibers that may not end, e.g. in [Signal](https://github.com/functional-streams-for-scala/fs2/blob/master/core/shared/src/main/scala/fs2/concurrent/Signal.scala#L169):

```scala
Concurrent[F].start(pubSub.publish((access._1, (access._1 + 1, a)))).as(true)
```

This code will undoubtedly leak the fiber and the queue in ZIO, if a queue is discarded before all such publishing fibers finish. Whether or not this exact code causes this exact leak, this code WILL leak eventually.

Clearly, there's a lot of code in the wild that assumes that Fibers, unlike Threads, are not garbage collection roots and it's completely OK to spawn non-terminating fibers and rely on GC to clean all non-referenced Fibers. I think this is a fundamental property of Fibers, too, and must be preserved.

I think it's untenable to use strong sets that make Fibers into GC roots directly or through their parents, the choice has been made for us. Now, JS does not support retrieval of keys from WeakMaps and WeakSets https://github.com/scala-js/scala-js/issues/3503, this was mentioned in https://github.com/zio/zio/issues/1081 and, I assume, is the principal reason for utilizing strong Set as the root set. I don't see any way to workaround that, so JS will just have to get second-class support here. On the JVM, though we'll just need to a WeakHashSet. There isn't a ConcurrentWeakHashSet implementation in the standard library but we can just steal one of them from e.g. [specs2](https://github.com/etorreborre/specs2/blob/8305db76c5084e4b3ce5827ce23117f6fb6beee4/common/shared/src/main/scala/org/specs2/control/eff/Cache.scala#L117) or write a new one.

**Comment by jdegoes** [comment]
@neko-kai I agree we will benefit from these changes and it seems like the most sensible approach (an alternative is to enhance `daemon` mode so you can specify whether or not to track the daemon in the root; if you applied such a mode to the top level main, you would get current behavior back). 

Unfortunately, it will make testing tricky because when a child is GC'd is non-deterministic.

I'll return to this as soon as I get some free time (maybe Sunday), if someone else wants to hack on it, then that's fine with me.

**Comment by sebver** [comment]
> @sebver Any luck with a reproduction yet?

Hi @neko-kai, unfortunately I haven't found the time yet for a reproduction, but will hopefully come around to that this weekend.

**Comment by Fristi** [comment]
@sebver Ran the app locally with `nonDaemon` and without. We dumped the heap in a hprof file and opened it with JProfiler on my machine. 

- On the left daemon mode
- On the right non-daemon mode

Both seem to allocate a lot of heap space by instantiating `zio.Fiber`

<img width="1917" alt="Screenshot 2019-12-04 at 10 00 46" src="https://user-images.githubusercontent.com/141265/70128480-2ecaff00-167d-11ea-9344-6e8d9ece1cf4.png">

**Comment by sebver** [comment]
I finally managed to make a reproduction, sorry for the long wait @neko-kai. The leak is actually only happening in combination with the http4s blaze client. All other libraries + zio we use work fine.

https://github.com/sebver/zio-memory-issue/

After a few mins it should be throwing an OOM. I also added a cats version and checked what happened with zio RC16, both don't seem to have the issue. Adding `nonDaemon` at the end didn't seem to do anything.

**Comment by joprice** [comment]
Maybe this is fixed? Bumping to the newest release showed constant 50mb sawtooth for me  https://github.com/sebver/zio-memory-issue/compare/master...joprice:master

**Comment by neko-kai** [comment]
@sebver @Fristi Could you confirm whether the issue is fixed on `RC-18-2` ?

**Comment by sebver** [comment]
Cool, I can confirm this issue is resolved now 👍

**Comment by Fristi** [comment]
I have no issues with RC18-2

**Comment by adamgfraser** [comment]
Closed by #2748.
